### PR TITLE
norman: volume-to-surface cross-attention for near-wall pressure coupling

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Injecting near-wall volume pressure tokens into the surface prediction head via a learned cross-attention module will improve both volume_pressure and τ_y/τ_z metrics by explicitly coupling the surface wall-shear model to the near-body pressure field.**

The current Transolver model processes surface points and volume points in **independent streams** that are only weakly coupled through the shared slice-attention backbone. This is physically incorrect: wall-shear stress at a surface point is determined by the pressure gradient normal to the wall — the volume pressure in the adjacent near-wall region drives the surface shear patterns directly. In RANS, the momentum equations enforce `∂p/∂x_wall ≈ ρ·u·∂u/∂y` at the wall, so near-wall `p_vol` and `τ_wall` are tightly coupled.

**The specific problem:** The τ_y/τ_z gap (2.35×/2.73× above AB-UPT) concentrates at separation lines and vortex footprints, where near-wall volume pressure has distinctive structure (adverse pressure gradient). The model currently predicts surface shear from surface geometry alone, without access to this local near-wall pressure signal.

**Proposed fix:** After the shared Transolver backbone produces per-token embeddings for both surface and volume streams, add a single cross-attention layer where **surface tokens attend to near-wall volume tokens**:

1. For each surface token, find the K=8 nearest volume tokens (by 3D distance, precomputed)
2. Apply multi-head cross-attention: queries = surface token embeddings, keys/values = corresponding near-wall volume token embeddings
3. Add the cross-attention output (residual) to the surface token before the surface prediction MLP head

This creates an explicit information path: `vol_pressure_gradient → surface_token → τ_prediction`. Total parameter overhead is ~4 attention heads × d_head = O(512·4) = 2M params.

**Why this hasn't been done before:** The Transolver architecture was adapted from a general PDE solver. Its default design assumes surface and volume are independent tasks. The physics of RANS wall boundary conditions says they're not. This is the first experiment to add an explicit cross-stream coupling.

Reference: cross-attention between mesh regions is standard in graph neural network CFD papers (e.g., MeshGraphNets, Li et al. 2023 "Geometry-Informed Neural Operator"), but has not been applied in our Transolver-based stack.

### Step 1 — Add CLI flags

```python
parser.add_argument('--vol-surf-crossattn', action='store_true', default=False,
    help='Add volume-to-surface cross-attention: surface tokens attend to K nearest volume tokens before the surface prediction head.')
parser.add_argument('--vol-surf-crossattn-k', type=int, default=8,
    help='Number of nearest volume tokens each surface token attends to (default 8).')
parser.add_argument('--vol-surf-crossattn-heads', type=int, default=4,
    help='Number of attention heads in the volume-to-surface cross-attention layer (default 4).')
```

### Step 2 — Implement the cross-attention module

Add this class to `target/train.py` or a small `crossattn.py` module (import from train.py):

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class VolToSurfCrossAttn(nn.Module):
    """
    Cross-attention: surface tokens (queries) attend to near-wall volume tokens (keys/values).
    Designed to inject near-wall pressure information into surface shear prediction.
    """
    def __init__(self, d_model: int, n_heads: int = 4, dropout: float = 0.0):
        super().__init__()
        assert d_model % n_heads == 0
        self.n_heads = n_heads
        self.d_head = d_model // n_heads
        self.q_proj = nn.Linear(d_model, d_model, bias=False)
        self.k_proj = nn.Linear(d_model, d_model, bias=False)
        self.v_proj = nn.Linear(d_model, d_model, bias=False)
        self.out_proj = nn.Linear(d_model, d_model, bias=False)
        self.norm_q = nn.LayerNorm(d_model)
        self.norm_kv = nn.LayerNorm(d_model)
        self.dropout = nn.Dropout(dropout)
        # Initialize output projection to near-zero so residual starts as identity
        nn.init.normal_(self.out_proj.weight, std=0.02 / (d_model ** 0.5))

    def forward(self, surf_emb, vol_emb, vol_nn_idx):
        """
        surf_emb:   [B, N_surf, D] — surface token embeddings from Transolver backbone
        vol_emb:    [B, N_vol,  D] — volume token embeddings from Transolver backbone
        vol_nn_idx: [B, N_surf, K] — precomputed K nearest volume token indices per surface token
        Returns:    [B, N_surf, D] — updated surface embeddings (residual-added)
        """
        B, N_surf, D = surf_emb.shape
        K = vol_nn_idx.shape[-1]

        # Normalize before attention (pre-norm style)
        q_in = self.norm_q(surf_emb)   # [B, N_surf, D]
        kv_in = self.norm_kv(vol_emb)  # [B, N_vol, D]

        # Gather K nearest volume token embeddings for each surface token
        idx = vol_nn_idx.unsqueeze(-1).expand(B, N_surf, K, D)   # [B, N_surf, K, D]
        kv_in_exp = kv_in.unsqueeze(2).expand(B, N_vol, 1, D)    # wrong shape — use gather
        # Correct gather:
        kv_gathered = torch.gather(
            kv_in.unsqueeze(1).expand(B, N_surf, N_vol, D),
            2,
            vol_nn_idx.unsqueeze(-1).expand(B, N_surf, K, D)
        )  # [B, N_surf, K, D] — this is memory-intensive; use the efficient version below

        # Memory-efficient gather (avoids N_surf × N_vol expansion):
        vol_nn_idx_flat = vol_nn_idx.reshape(B, -1)  # [B, N_surf*K]
        kv_gathered_flat = torch.gather(
            kv_in,
            1,
            vol_nn_idx_flat.unsqueeze(-1).expand(B, N_surf * K, D)
        )  # [B, N_surf*K, D]
        kv_gathered = kv_gathered_flat.reshape(B, N_surf, K, D)   # [B, N_surf, K, D]

        # Project to Q, K, V
        Q = self.q_proj(q_in)                          # [B, N_surf, D]
        K_ = self.k_proj(kv_gathered)                  # [B, N_surf, K, D]
        V  = self.v_proj(kv_gathered)                  # [B, N_surf, K, D]

        # Reshape for multi-head attention
        Q = Q.reshape(B, N_surf, self.n_heads, self.d_head).transpose(1, 2)
        # [B, H, N_surf, d_head]
        K_ = K_.reshape(B, N_surf, K, self.n_heads, self.d_head).permute(0, 3, 1, 4, 2)
        # [B, H, N_surf, d_head, K]
        V  = V.reshape(B, N_surf, K, self.n_heads, self.d_head).permute(0, 3, 1, 2, 4)
        # [B, H, N_surf, K, d_head]

        # Scaled dot-product attention over K neighbors
        scale = self.d_head ** -0.5
        attn = (Q.unsqueeze(-1) * K_).sum(-2) * scale  # [B, H, N_surf, K] — dot product
        attn = F.softmax(attn, dim=-1)                  # normalize over K neighbors
        attn = self.dropout(attn)

        # Weighted sum of values
        out = (attn.unsqueeze(-1) * V).sum(-2)          # [B, H, N_surf, d_head]
        out = out.transpose(1, 2).reshape(B, N_surf, D)  # [B, N_surf, D]

        # Output projection + residual
        out = self.out_proj(out)
        return surf_emb + out
```

### Step 3 — Precompute nearest volume indices per surface point

Before training, compute the K nearest volume tokens for each surface token (once per batch, since surface/volume point sets change per batch via random sampling):

```python
# Inside the training step, after obtaining surface_coords [B, N_surf, 3] and vol_coords [B, N_vol, 3]:
if args.vol_surf_crossattn:
    with torch.no_grad():
        # Compute pairwise distances: [B, N_surf, N_vol]
        # Efficient via cdist (avoids explicit N^2 materialization for large N)
        # With N_surf=N_vol=65536 this is 65536^2 = 4GB per batch — TOO LARGE
        # Use approximate kNN or chunked distance:
        K = args.vol_surf_crossattn_k
        chunk_size = 1024  # process 1024 surface points at a time
        vol_nn_idx = torch.zeros(B, N_surf, K, dtype=torch.long, device=device)
        for chunk_start in range(0, N_surf, chunk_size):
            chunk_end = min(chunk_start + chunk_size, N_surf)
            surf_chunk = surface_coords[:, chunk_start:chunk_end, :]  # [B, C, 3]
            # [B, C, N_vol] pairwise L2
            dists = torch.cdist(surf_chunk, vol_coords)               # [B, C, N_vol]
            vol_nn_idx[:, chunk_start:chunk_end, :] = dists.topk(K, dim=-1, largest=False).indices
```

**Performance note:** With N_surf=N_vol=65536, full cdist is too large. You must use the chunked version above (chunk_size=1024 → 64 chunks × [B, 1024, 65536] = manageable). Report the per-step overhead in your first comment (should be < 5 seconds/step total for K=8).

Alternative: use a sparser surface/volume (e.g. N=16384) for the cross-attention while keeping 65536 for the main model — but try the chunked cdist first.

### Step 4 — Wire into the model forward pass

After the Transolver backbone produces `surf_emb [B, N_surf, D]` and `vol_emb [B, N_vol, D]`, apply the cross-attention before the surface head:

```python
# After backbone, before surface prediction MLP:
if args.vol_surf_crossattn:
    surf_emb = vol_surf_crossattn_module(surf_emb, vol_emb, vol_nn_idx)
# Then pass surf_emb to surface prediction MLP as usual
```

Instantiate the module once in model initialization:
```python
if args.vol_surf_crossattn:
    self.vol_surf_crossattn = VolToSurfCrossAttn(
        d_model=args.model_hidden_dim,
        n_heads=args.vol_surf_crossattn_heads
    )
```

### Step 5 — Diagnostics

Log per step:
- `train/crossattn_attn_entropy` — mean entropy of attention distribution over K neighbors (if near log(K)=2.08 → uniform, not learning; if near 0 → sharp locality)
- `train/crossattn_out_norm` — L2 norm of the cross-attention output before residual (should start small due to near-zero init and grow as training proceeds)

At epoch end, report per-channel val metrics (surface_pressure, tau_y, tau_z, volume_pressure).

### Step 6 — Sweep (2 arms, sequential on 4×H100 DDP)

**Arm A (control):** `--no-vol-surf-crossattn` (standard independent streams)  
**Arm B:** `--vol-surf-crossattn` (4 cross-attention heads, K=8 nearest vol tokens)

Use `--wandb-group norman-r34-vol-surf-xattn` for both arms.

**Full launch command for Arm A (control):**
```bash
cd target/
nohup torchrun --standalone --nproc_per_node=4 train.py \
  --agent norman \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --clip-grad-norm 0.5 \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --learnable-pe \
  --no-compile-model \
  --epochs 3 \
  --wandb-group norman-r34-vol-surf-xattn \
  --wandb-name norman-r34-arm-a-ctrl \
  > _runs/arm_a_ctrl.log 2>&1 &
echo "PID: $!"
```

For Arm B add: `--vol-surf-crossattn --vol-surf-crossattn-k 8 --vol-surf-crossattn-heads 4`

### IMPORTANT

- Use `--clip-grad-norm 0.5` (NOT `--grad-clip-norm`). Verify: `python train.py --help | grep -i clip`
- If the chunked cdist is > 10s/step, reduce surface/volume points to 16384 for the kNN computation (subsample coordinates before cdist, gather embeddings at the same indices)
- Run Arm A first — verify it matches ~9.0% val_abupt at epoch 3 (confirms your setup is correct) before launching Arm B

## Results to report

For each arm:
1. W&B run id + link
2. Per-step overhead of kNN computation (seconds, Arm B only)
3. Per-epoch `val/abupt_axis_mean_rel_l2_pct` table
4. Per-channel at best epoch: `surface_pressure`, `tau_y`, `tau_z`, `volume_pressure`
5. `train/crossattn_attn_entropy` at epochs 1, 2, 3 (did it sharpen from log(8)=2.08?)
6. `test/abupt_axis_mean_rel_l2_pct` from best-val checkpoint

Win criterion: **val_abupt < 9.032%** AND volume_pressure OR τ_y/τ_z shows improvement vs Arm A. If attn_entropy stays near log(8) (uniform attention over neighbors), the cross-attention is not learning a useful locality signal — report this explicitly.

## Baseline

Current best on yi:
- **val_abupt: 9.032%** (PR #517 askeladd, W&B run `brat65z4`, Lion lr=1e-4 clip=0.5)
- STRING-sep learnable PE now on yi via PR #490; frieren resumed run reached **8.087%** with `--learnable-pe` (awaiting clean from-scratch confirmation from PR #539)
- Expected next bar: ~8.1–8.5% once PR #539 lands

AB-UPT reference targets:
- τ_y (ws_y): 3.65% — **current yi: ~8.58%, gap 2.35×**
- τ_z (ws_z): 3.63% — **current yi: ~9.93%, gap 2.73×**
- volume_pressure: 6.08% — **current yi: ~11.87%, gap 1.95×**
- surface_pressure: 3.82%

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT norman:

## Status update

**Arm A control is running** (PID 320693, started before this session). At 02:40 UTC it's at epoch 1, 35% (1887/5442 steps, 1.25 s/it). Run config matches PR instructions exactly: Lion lr=1e-4 clip=0.5 batch=4 4L/512d/8h/128sl learnable-pe ema=0.999 65536 surface/volume points 3 epochs DDP world_size=4. ETA all 3 epochs ~07:40 UTC.

W&B: norman-r34-arm-a-ctrl run id 2xpbh4nx (group: norman-r34-vol-surf-xattn).

## Implementation review

Implementation already in place from earlier session. I verified correctness via a focused Python smoke test:
- kNN respects `volume_mask` (no masked-region indices selected)
- Cross-attn output equals input at init (zero-init out_proj works)
- attn_entropy = log(8) = 2.079 at init (perfect uniform)
- attn_max = 0.125 at init (perfect uniform)
- out_norm = 0.0 at init
- Backward pass produces finite grads

Two refinements made to the existing diff:
1. Fixed a wandb double-prefix bug: diagnostics keys were `crossattn/crossattn_attn_entropy` and `crossattn/crossattn_out_norm`. Renamed inner keys to `attn_entropy` / `out_norm` (final wandb names: `train/crossattn/attn_entropy`, `train/crossattn/out_norm`).
2. Added a third diagnostic `attn_max` (mean of per-token max softmax weight; near 1/K=0.125 → uniform/non-discriminating, > 0.5 → peaked/learned locality).

## Plan

Once Arm A completes its 3 epochs and frees the 4 GPUs, launch Arm B (`--vol-surf-crossattn --vol-surf-crossattn-k 8 --vol-surf-crossattn-heads 4`) on the same hardware. Will report results from both arms once Arm B finishes.

**Comment by morganmcg1** [comment]
ADVISOR: Closing this PR — a previous advisor loop already added a closing comment but the PR was not formally closed. All of these approaches were reviewed and did not beat the current baseline (val_abupt 9.032%). Freeing this student for a new assignment.
